kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/EnvelopeEncryption.java

```diff
@@ -69,11 +70,17 @@ public Config initialize(FilterFactoryContext context, Config config) throws Plu
     public EnvelopeEncryptionFilter<K> createFilter(FilterFactoryContext context, Config configuration) {
         Kms<K, E> kms = buildKms(context, configuration);
 
-        var keyManager = new InBandKeyManager<>(kms, BufferPool.allocating(), 500_000);
+        DekManager<K, E> dekManager = new DekManager<>(ignored -> kms, null, 5_000_000);
+        var encryptionManager = new InBandEncryptionManager<>(dekManager,
+                1024 * 1024,
+                8 * 1024 * 1024,
```

Comment:
Absolutely, but I don't think we need to do that in this PR. Ideally we'd do the work to do the per-thread allocation and surface them then, because at that point we're giving the user config parameters which they can reason much more usefully about, in particular "I have T threads and B bytes of buffer per thread, so I'm using T×B bytes for these buffers in total".

FWIW, last night you speculated whether we really needed per-thread, or whether per-channel was enough. I think we do need per-thread. Per channel doesn't give the user any config parameters to actually bound total memory usage for these buffers (there's no bound on the number of channels). I think that's important because I can imagine with compression and enough connections using per-channel would make you more vulnerable to OOME.

Reply:
Also on that topic I think I heard a question about whether it should be per-event-loop rather than per-thread.

A netty event loop is single threaded, I don't think there's an explicit document stating that it's guaranteed for all EventLoop implementations, but all the netty 4 EventLoop implementations extend `SingleThreadEventLoop` (except EmbeddedEventLoop which is deterministic and driven by a client, not a thread). There is also a deprecated ThreadPerChannelEventLoop for if you are running with old-io, the guarantee is still 1 thread per channel, but a lot more threads. We don't use this old implementation so we can rely on the idea of 1 channel -> 1 eventloop -> 1 thread with a bounded number of threads in play.

Reply:
> edit: one concern is making sure we are on the eventloop when we need to be, after any KMS operations have completed

Yes, but we want that guarantee anyway, I think, otherwise we need to start worrying about races between KMS threads and event loop threads for all state, not just these buffers.

Reply:
Yeah certainly happy to expose configuration in a separate PR.

> "I have T threads and B bytes of buffer per thread, so I'm using T×B bytes for these buffers in total"

I hear the point your making but we don't allow users to con figure/tune `T`, I'm not even sure we log `T` so they still can't reason about TxB. But if thats our goal should we even expose the buffer sizes? We know the memory from XMS/cgroups limits and we know the thread count so maybe we just divide the memory space using some heuristic for proxy overheads?

Reply:
> But if thats our goal should we even expose the buffer sizes?

Maybe not, but that's precisely what we should figure out in the later PR.